### PR TITLE
Terraform Module for EFS

### PR DESCRIPTION
**Pull request by angeleenasunny**
Terraform Plan:
-----------------------
  
  # module.efs.aws_efs_file_system.elastic_file_system will be created
  + resource "aws_efs_file_system" "elastic_file_system" {
      + arn                     = (known after apply)
      + availability_zone_id    = (known after apply)
      + availability_zone_name  = (known after apply)
      + creation_token          = "test-efs"
      + dns_name                = (known after apply)
      + encrypted               = true
      + id                      = (known after apply)
      + kms_key_id              = (known after apply)
      + name                    = (known after apply)
      + number_of_mount_targets = (known after apply)
      + owner_id                = (known after apply)
      + performance_mode        = "generalPurpose"
      + size_in_bytes           = (known after apply)
      + tags_all                = {
          + "Environment" = "Test"
          + "Project"     = "QBurst"
        }
      + throughput_mode         = "bursting"

      + lifecycle_policy {
          + transition_to_ia = "AFTER_30_DAYS"
        }
    }

  # module.efs.aws_efs_mount_target.efs_mount_target[0] will be created
  + resource "aws_efs_mount_target" "efs_mount_target" {
      + availability_zone_id   = (known after apply)
      + availability_zone_name = (known after apply)
      + dns_name               = (known after apply)
      + file_system_arn        = (known after apply)
      + file_system_id         = (known after apply)
      + id                     = (known after apply)
      + ip_address             = (known after apply)
      + mount_target_dns_name  = (known after apply)
      + network_interface_id   = (known after apply)
      + owner_id               = (known after apply)
      + security_groups        = [
          + "sg-01234567890abcdef",
        ]
      + subnet_id              = "subnet-01234567890abcdef"
    }

  # module.efs.aws_efs_mount_target.efs_mount_target[1] will be created
  + resource "aws_efs_mount_target" "efs_mount_target" {
      + availability_zone_id   = (known after apply)
      + availability_zone_name = (known after apply)
      + dns_name               = (known after apply)
      + file_system_arn        = (known after apply)
      + file_system_id         = (known after apply)
      + id                     = (known after apply)
      + ip_address             = (known after apply)
      + mount_target_dns_name  = (known after apply)
      + network_interface_id   = (known after apply)
      + owner_id               = (known after apply)
      + security_groups        = [
          + "sg-01234567890abcdef",
        ]
      + subnet_id              = "subnet-01234567890ghijklm"
    }

  # module.efs.aws_kms_key.kms_key_efs will be created
  + resource "aws_kms_key" "kms_key_efs" {
      + arn                                = (known after apply)
      + bypass_policy_lockout_safety_check = false
      + customer_master_key_spec           = "SYMMETRIC_DEFAULT"
      + description                        = "KMS key for EFS"
      + enable_key_rotation                = true
      + id                                 = (known after apply)
      + is_enabled                         = true
      + key_id                             = (known after apply)
      + key_usage                          = "ENCRYPT_DECRYPT"
      + multi_region                       = (known after apply)
      + policy                             = (known after apply)
      + tags_all                           = {
          + "Environment" = "Test"
          + "Project"     = "QBurst"
        }
    }

Plan: 4 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + example_outputs = {
      + efs_arn      = (known after apply)
      + efs_id       = (known after apply)
      + efs_mount_id = [
          + (known after apply),
          + (known after apply),
        ]
    }

### CODE REVIEW

File: terraform/aws/modules/efs/efs.tf [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/*This will create EFS in bursting mode.The mode can be changed as per the requirement*/
+resource "aws_efs_file_system" "elastic_file_system" {
+  creation_token = var.creation_token
+
+  lifecycle_policy {
+    transition_to_ia = "AFTER_30_DAYS"
+  }
```
Comment (qburst-praven) [inline terraform/aws/modules/efs/efs.tf:7]: pls add options to set encryption true/false as needed and the kms_key_id as well.

File: terraform/aws/modules/efs/efs.tf [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  performance_mode = var.performance_mode
+  throughput_mode  = var.throughput_mode
+}
+
```
Comment (qburst-praven) [inline terraform/aws/modules/efs/efs.tf:11]: Please add an optional EFS mount id as well.

**Review by qburst-praven — CHANGES_REQUESTED** [review]

File: terraform/aws/modules/efs/efs.tf (hunk 1 above)
Comment (angeleenasunny) [reply terraform/aws/modules/efs/efs.tf:7]: @qburst-praven This is done.

File: terraform/aws/modules/efs/efs.tf (hunk 2 above)
Comment (angeleenasunny) [reply terraform/aws/modules/efs/efs.tf:13]: @qburst-praven Created a mount target and terraform plan is updated in the PR comment

File: terraform/aws/modules/efs/variables.tf [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+variable "throughput_mode" {
+  type    = string
+  default = "bursting"
+
+}
+variable "encryption_set" {
+  type    = bool
+  default = true
+}
+variable "kms_key_id" {
```
Comment (qburst-praven) [inline terraform/aws/modules/efs/variables.tf:20]: @angeleenasunny , I believe this variable is not needed anymore?

File: terraform/aws/modules/efs/efs.tf [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  }
+  encrypted        = var.encryption_set
+  kms_key_id       = aws_kms_key.kms_key_efs.arn
+  performance_mode = var.performance_mode
+  throughput_mode  = var.throughput_mode
+}
+
+# Define the EFS mount target (optional)
+resource "aws_efs_mount_target" "efs_mount_target" {
+  file_system_id  = aws_efs_file_system.elastic_file_system.id
```
Comment (qburst-praven) [inline terraform/aws/modules/efs/efs.tf:16]: please add a count logic. Only create the mount target, if the subnet is provided. Additionally, you can change subnet to be a list, that way for each subnet provided you can create a new mount target. This is useful in cases of multi AZ.

File: terraform/aws/modules/efs/variables.tf [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  type    = string
+  default = "true"
+
+}
+variable "subnet_id" {
+  type    = string
+  default = "subnet-01234567890abcdef"
```
Comment (qburst-praven) [inline terraform/aws/modules/efs/variables.tf:27]: pls change the default values, these and the sec groups provided below might not exist.

**Review by qburst-praven — CHANGES_REQUESTED** [review]

File: terraform/aws/modules/efs/variables.tf (hunk 3 above)
Comment (angeleenasunny) [reply terraform/aws/modules/efs/variables.tf]: @qburst-praven This is removed.

File: terraform/aws/modules/efs/efs.tf (hunk 4 above)
Comment (angeleenasunny) [reply terraform/aws/modules/efs/efs.tf:17]: @qburst-praven This is done.

**Review by qburst-praven — APPROVED** [review]

**Pull request merged**